### PR TITLE
(LTH-54) Formalize library installs

### PR DESCRIPTION
**Pull request by MikaelSmith**
Create a `leatherman_install` macro to replace CMake's `install` with
one supporting cross-platform defaults.

### CODE REVIEW

**Comment by whopper** [comment]
Ping @branan for review - this seems like a pretty light, sane change, but cmake is still pretty new to me so I wouldn't be able to spot any subtle issues in this -- so calling on your cmake knowledge for a quick review and merge :)

File: cmake/leatherman.cmake [hunk 1]
```diff
@@ -59,6 +59,26 @@ macro(defoption name doc default)
...
+# On Windows shared libraries go in bin, import and archive libraries
+# go in lib. On Linux shared libraries go in lib. Binaries go in bin.
+#
+# Also always drop the prefix; give the target its expected name.
+# We often have binaries and related dynamic libraries, and this
+# simplifies giving them different but related names, such as
+# `facter` and `libfacter`.
+macro(leatherman_install)
+    install(TARGETS ${ARGV}
+        RUNTIME DESTINATION bin
```
Comment (branan) [inline cmake/leatherman.cmake]: I think these RUNTIME, LIBRARY, and ARCHIVE destinations are all the CMake defaults?

File: cmake/leatherman.cmake [hunk 2]
```diff
@@ -147,3 +167,19 @@ function(get_commit_string varname)
...
+
+include(GenerateExportHeader)
+# Usage: symbol_exports(TARGET HEADER)
+#
+# Generate the export header for restricting symbols exported from the library,
+# and configure the compiler. Restricting symbols has several advantages, noted
+# at https://gcc.gnu.org/wiki/Visibility.
+macro(symbol_exports target header)
+    generate_export_header(${target} EXPORT_FILE_NAME "${header}")
+    if (NOT APPLE)
```
Comment (branan) [inline cmake/leatherman.cmake]: Why is this "NOT APPLE"? Do we also need to constrain output stuff on other unices besides linux (IIRC AIX doesn't support visibility?)

Comment (MikaelSmith) [inline cmake/leatherman.cmake]: AIX will safely ignore it. Apple will try to use it, but then fail to export things we need so the libraries are unusable. There were some hangups we haven't figured out.

This worked at https://github.com/puppetlabs/cpp-pcp-client/blob/master/lib/CMakeLists.txt#L41-L50, so it should be safe for all our current platforms.

Comment (branan) [inline cmake/leatherman.cmake]: Makes sense, thanks

File: cmake/leatherman.cmake (hunk 1 above)
Comment (MikaelSmith) [inline cmake/leatherman.cmake]: Worth checking - I'm not sure import libraries go where we want on Windows.

Comment (MikaelSmith) [inline cmake/leatherman.cmake]: It might be useful to update it to support https://github.com/puppetlabs/facter/pull/1228.

Comment (MikaelSmith) [inline cmake/leatherman.cmake]: If I remove all destinations, there don't appear to be any defaults:
`install TARGETS given no ARCHIVE DESTINATION for static library target "leatherman_logging".`

**Comment by MikaelSmith** [comment]
Depends on https://github.com/puppetlabs/leatherman/pull/107 for shared library installs to work.

**Comment by MikaelSmith** [comment]
Should be fixed up.

File: cmake/leatherman.cmake (hunk 2 above)
Comment (MikaelSmith) [inline cmake/leatherman.cmake]: I'll add a comment.

**Comment by MikaelSmith** [comment]
ping @branan

**Pull request merged**